File: ios/Runner/RemindersBridge.swift
import EventKit
import Flutter
import Foundation

/// MethodChannel bridge for the iOS Reminders framework. Mirrors
/// the `agent_buddy/reminders` protocol. iOS gives us a real
/// to-do store via `EKReminder`, so unlike Android we don't
/// piggy-back on calendars.
///
/// Permission flow mirrors `CalendarBridge` and `LocationBridge`:
/// every concurrent call is parked in a per-call pending map so
/// the user's tap on the system dialog resumes **all** of them —
/// not just the most recent one. This avoids the "tool call fails
/// before the user can answer" race when the model fires multiple
/// tool calls in quick succession.
public class RemindersBridge: NSObject, FlutterPlugin {
  public static let channelName = "agent_buddy/reminders"

  private let store = EKEventStore()
  private var pendingByCallId: [String: Pending] = [:]
  private var nextCallSeq: Int = 0
  private let stateLock = NSLock()
  private var requestInFlight: Bool = false

  private struct Pending {
    let result: FlutterResult
    let method: String
    let args: [String: Any]
    let fireLock: NSLock
    var delivered: Bool
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    let instance = RemindersBridge()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult
  ) {
    let args = call.arguments as? [String: Any] ?? [:]
    switch call.method {
    case "ensurePermission":
      // Pure read; respond synchronously.
      result(["granted": store.authorizationStatus(for: .reminder) == .fullAccess])
    case "listCalendars":
      // iOS doesn't expose the "pick a calendar" picker flow that
      // Android needs. We still implement the method so the Dart
      // side gets a non-empty list back; the picker sheet is a
      // no-op on iOS because the system Reminders app handles
      // the default list.
      result(self.listWritableCalendars())
    case "setTodoCalendar", "getTodoCalendar":
      // No-op on iOS: the system Reminders framework has a single
      // canonical store. The Dart side knows this and short-
      // circuits the picker sheet.
      if call.method == "getTodoCalendar" {
        result("ios_default")
      } else {
        result(["ok": true])
      }
    case "listReminders":
      parkOrDispatch(call, result: result, method: call.method, args: args)
    case "createReminder":
      parkOrDispatch(call, result: result, method: call.method, args: args)
    case "completeReminder":
      parkOrDispatch(call, result: result, method: call.method, args: args)
    case "updateReminder":
      parkOrDispatch(call, result: result, method: call.method, args: args)
    case "deleteReminder":
      parkOrDispatch(call, result: result, method: call.method, args: args)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Permission dispatcher

  private func parkOrDispatch(
    _ call: FlutterMethodCall,
    result: @escaping FlutterResult,
    method: String,
    args: [String: Any]
  ) {
    let status = store.authorizationStatus(for: .reminder)
    if status == .fullAccess {
      DispatchQueue.main.async { [weak self] in
        self?.dispatch(method: method, args: args, originalResult: result)
      }
      return
    }
    if status == .denied || status == .restricted {
      // The user has permanently denied access; don't pop another
      // dialog. Surface PERMANENTLY_DENIED so the AI can hint at
      // system settings.
      result(
        FlutterError(
          code: "PERMANENTLY_DENIED",
          message: "Reminders permission permanently denied; open system settings to enable it",
          details: nil
        )
      )
      return
    }
    // .notDetermined: ask. Park the call.
    let callId = registerPending(result: result, method: method, args: args)
    requestAccessIfNeeded(callId: callId)
  }

  private func registerPending(
    result: @escaping FlutterResult,
    method: String,
    args: [String: Any]
  ) -> String {
    stateLock.lock()
    nextCallSeq += 1
    let callId = "rem-\(nextCallSeq)"
    pendingByCallId[callId] = Pending(
      result: result,
      method: method,
      args: args,
      fireLock: NSLock(),
      delivered: false
    )
    stateLock.unlock()
    return callId
  }

  private func requestAccessIfNeeded(callId: String) {
    stateLock.lock()
    let inFlight = requestInFlight
    if !inFlight { requestInFlight = true }
    stateLock.unlock()
    if inFlight { return }

    if #available(iOS 17.0, *) {
      Task { @MainActor in
        _ = try? await self.store.requestFullAccessToReminders()
        DispatchQueue.main.async { [weak self] in
          self?.resumeAllPending()
        }
      }
    } else {
      self.store.requestAccess(to: .reminder) { [weak self] _, _ in
        DispatchQueue.main.async {
          self?.resumeAllPending()
        }
      }
    }
  }

  private func resumeAllPending() {
    stateLock.lock()
    requestInFlight = false
    let status = self.store.authorizationStatus(for: .reminder)
    let snapshot = self.pendingByCallId
    stateLock.unlock()

    switch status {
    case .fullAccess:
      for (cid, _) in snapshot {
        self.deliverResume(cid)
      }
    case .denied, .restricted:
      for (cid, _) in snapshot {
        self.deliverError(
          cid,
          code: "PERMANENTLY_DENIED",
          message: "Reminders permission permanently denied; open system settings to enable it"
        )
      }
    case .notDetermined:
      for (cid, _) in snapshot {
        self.deliverError(
          cid,
          code: "PERMISSION_DENIED",
          message: "Reminders permission dialog was dismissed without a decision."
        )
      }
    @unknown default:
      for (cid, _) in snapshot {
        self.deliverError(
          cid,
          code: "PERMISSION_DENIED",
          message: "Reminders permission state is unknown."
        )
      }
    }
  }

  private func deliverResume(_ callId: String) {
    stateLock.lock()
    guard var p = pendingByCallId[callId] else {
      stateLock.unlock()
      return
    }
    p.fireLock.lock()
    if p.delivered {
      p.fireLock.unlock()
      stateLock.unlock()
      return
    }
    p.delivered = true
    p.fireLock.unlock()
    pendingByCallId.removeValue(forKey: callId)
    stateLock.unlock()
    dispatch(method: p.method, args: p.args, originalResult: p.result)
  }

  private func deliverError(_ callId: String, code: String, message: String) {
    stateLock.lock()
    guard var p = pendingByCallId[callId] else {
      stateLock.unlock()
      return
    }
    p.fireLock.lock()
    if p.delivered {
      p.fireLock.unlock()
      stateLock.unlock()
      return
    }
    p.delivered = true
    p.fireLock.unlock()
    pendingByCallId.removeValue(forKey: callId)
    stateLock.unlock()
    p.result(FlutterError(code: code, message: message, details: nil))
  }

  // MARK: - Dispatch (already authorized path)

  private func dispatch(
    method: String,
    args: [String: Any],
    originalResult: @escaping FlutterResult
  ) {
    switch method {
    case "listReminders":
      let includeCompleted = args["includeCompleted"] as? Bool ?? false
      let max = (args["max"] as? NSNumber)?.intValue ?? 50
      // fetchReminders invokes its callback synchronously on the
      // calling thread, so we hop to a background queue to avoid
      // blocking the main thread.
      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        do {
          let reminders = try self.listReminders(
            includeCompleted: includeCompleted,
            max: max
          )
          DispatchQueue.main.async { originalResult(reminders) }
        } catch let err as BridgeError {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: err.code, message: err.message, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: "BRIDGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "createReminder":
      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        do {
          let r = try self.createReminder(args: args)
          DispatchQueue.main.async { originalResult(r) }
        } catch let err as BridgeError {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: err.code, message: err.message, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: "BRIDGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "completeReminder":
      let id = args["id"] as? String ?? ""
      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        do {
          let r = try self.completeReminder(id: id)
          DispatchQueue.main.async { originalResult(r as Any) }
        } catch let err as BridgeError {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: err.code, message: err.message, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: "BRIDGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "updateReminder":
      let id = args["id"] as? String ?? ""
      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        do {
          let r = try self.updateReminder(id: id, args: args)
          DispatchQueue.main.async { originalResult(r as Any) }
        } catch let err as BridgeError {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: err.code, message: err.message, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: "BRIDGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "deleteReminder":
      let id = args["id"] as? String ?? ""
      DispatchQueue.global(qos: .userInitiated).async { [weak self] in
        guard let self = self else { return }
        do {
          let ok = try self.deleteReminder(id: id)
          DispatchQueue.main.async { originalResult(["ok": ok]) }
        } catch let err as BridgeError {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: err.code, message: err.message, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            originalResult(FlutterError(code: "BRIDGE_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    default:
      originalResult(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Calendar listing (iOS shim)

  private func listWritableCalendars() -> [[String: Any]] {
    let cals = store.calendars(for: .reminder)
    return cals.map { cal in
      [
        "id": cal.calendarIdentifier,
        "displayName": cal.title,
        "accountName": cal.source?.title ?? "",
      ]
    }
  }

  // MARK: - Reminder CRUD

  private func listReminders(includeCompleted: Bool, max: Int) throws -> [[String: Any]] {
    let cals = store.calendars(for: .reminder)
    let predicate = store.predicateForIncompleteReminders(
      withDueDateStarting: nil,
      ending: nil,
      calendars: cals
    )
    // Caller is responsible for hopping off the main thread before
    // calling us; fetchReminders invokes its callback synchronously.
    var mapped: [[String: Any]] = []
    let semaphore = DispatchSemaphore(value: 0)
    store.fetchReminders(matching: predicate) { ekReminders in
      let filtered = (ekReminders ?? []).filter { includeCompleted || !$0.isCompleted }
      mapped = filtered.prefix(max).map { self.reminderToDict($0) }
      semaphore.signal()
    }
    if semaphore.wait(timeout: .now() + 10) == .timedOut {
      throw BridgeError(
        code: "BRIDGE_ERROR",
        message: "fetchReminders timed out"
      )
    }
    return mapped
  }

  private func createReminder(args: [String: Any]) throws -> [String: Any] {
    let title = args["title"] as? String ?? ""
    if title.isEmpty {
      throw BridgeError(code: "INVALID_ARGUMENT", message: "title required")
    }
    let notes = args["notes"] as? String
    let dueMs = (args["dueMs"] as? NSNumber)?.int64Value
    let reminder = EKReminder(eventStore: store)
    reminder.title = title
    reminder.notes = notes
    reminder.calendar = store.defaultCalendarForNewReminders()
    if let dueMs = dueMs {
      let date = Date(timeIntervalSince1970: TimeInterval(dueMs) / 1000.0)
      let comps = Calendar.current.dateComponents(
        [.year, .month, .day, .hour, .minute, .second],
        from: date
      )
      reminder.dueDateComponents = comps
    }
    do {
      try store.save(reminder, commit: true)
    } catch {
      throw BridgeError(
        code: "BRIDGE_ERROR",
        message: "failed to save reminder: \(error.localizedDescription)"
      )
    }
    return reminderToDict(reminder)
  }

  private func completeReminder(id: String) throws -> [String: Any]? {
    guard let calItem = store.calendarItem(withIdentifier: id) as? EKReminder else {
      return nil
    }
    calItem.isCompleted = true
    do {
      try store.save(calItem, commit: true)
    } catch {
      throw BridgeError(
        code: "BRIDGE_ERROR",
        message: "failed to complete reminder: \(error.localizedDescription)"
      )
    }
    return reminderToDict(calItem)
  }

  private func updateReminder(id: String, args: [String: Any]) throws -> [String: Any]? {
    guard let calItem = store.calendarItem(withIdentifier: id) as? EKReminder else {
      return nil
    }
    if let title = args["title"] as? String { calItem.title = title }
    if let notes = args["notes"] as? String { calItem.notes = notes }
    if let dueMs = (args["dueMs"] as? NSNumber)?.int64Value {
      let date = Date(timeIntervalSince1970: TimeInterval(dueMs) / 1000.0)
      let comps = Calendar.current.dateComponents(
        [.year, .month, .day, .hour, .minute, .second],
        from: date
      )
      calItem.dueDateComponents = comps
    }
    do {
      try store.save(calItem, commit: true)
    } catch {
      throw BridgeError(
        code: "BRIDGE_ERROR",
        message: "failed to update reminder: \(error.localizedDescription)"
      )
    }
    return reminderToDict(calItem)
  }

  private func deleteReminder(id: String) throws -> Bool {
    guard let calItem = store.calendarItem(withIdentifier: id) as? EKReminder else {
      return false
    }
    do {
      try store.remove(calItem, commit: true)
    } catch {
      throw BridgeError(
        code: "BRIDGE_ERROR",
        message: "failed to delete reminder: \(error.localizedDescription)"
      )
    }
    return true
  }

  // MARK: - Mapping

  private func reminderToDict(_ r: EKReminder) -> [String: Any] {
    var dict: [String: Any] = [
      "id": r.calendarItemIdentifier,
      "title": r.title ?? "",
      "notes": r.notes as Any,
      "completed": r.isCompleted,
      "completedAtMs": nil as Any?,
    ]
    if let dc = r.dueDateComponents,
       let date = Calendar.current.date(from: dc) {
      dict["dueMs"] = Int64(date.timeIntervalSince1970 * 1000)
    } else {
      dict["dueMs"] = nil as Any?
    }
    if r.isCompleted, let completed = r.completionDate {
      dict["completedAtMs"] = Int64(completed.timeIntervalSince1970 * 1000)
    }
    return dict
  }
}

File: android/app/src/main/kotlin/cn/leo/agent_buddy/RemindersBridge.kt
package cn.leo.agent_buddy

import android.Manifest
import android.app.Activity
import android.content.ContentUris
import android.content.ContentValues
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.provider.CalendarContract
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

/**
 * MethodChannel bridge for the "reminders / to-do" surface on
 * Android. Because Android has no unified system to-do API, we
 * piggy-back on the CalendarProvider: reminders are stored as
 * all-day events in a user-picked "todo" calendar. Each reminder
 * gets a marker prefix on its description so the list query stays
 * specific to Agent Buddy-created items.
 *
 * Channel: `agent_buddy/reminders`
 *
 * The user must pick a "todo calendar" the first time this tool is
 * used; until then, every call returns `NO_TODO_CALENDAR` so the
 * Dart side can prompt them via the picker sheet.
 *
 * Permission flow mirrors [CalendarBridge] / [LocationBridge]:
 * concurrent tool calls are all parked in a per-call pending map so
 * the user's tap on the system dialog resumes every one of them.
 *
 * Why a description-prefix marker instead of `categories`:
 * `CalendarContract.Events.CATEGORIES` is part of the contract but
 * several OEM calendar providers (notably Huawei / Xiaomi local
 * calendars) don't expose the column at all, so writing
 * `put("categories", ...)` blows up with
 * `IllegalArgumentException: column 'categories' is invalid`. The
 * `DESCRIPTION` column is universal, so we encode ownership as a
 * short sentinel prefix and strip it on read.
 */
class RemindersBridge(
    private val context: Context,
    private val activity: Activity,
) : MethodChannel.MethodCallHandler, PluginRegistry.RequestPermissionsResultListener {

    companion object {
        const val CHANNEL = "agent_buddy/reminders"
        private const val PREFS_NAME = "agent_buddy_prefs"
        private const val PREF_TODO_CALENDAR = "todo_calendar_id"
        private const val PERMISSION_REQUEST_CODE = 9002

        // Description prefix that marks an all-day event as an
        // Agent Buddy reminder. The trailing `|` is a separator so
        // the marker's length is well-defined when we strip it. The
        // SQLite LIKE wildcard `%` is appended in the SELECT
        // so `LIKE '__ab_agenda__|%'` only matches events we own.
        private const val MARKER = "__ab_agenda__|"
    }

    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private data class Pending(
        val result: MethodChannel.Result,
        val method: String,
        val args: Map<String, Any?>?,
        val delivered: AtomicBoolean,
    )

    private val pendingResults: MutableMap<String, Pending> = ConcurrentHashMap()
    private val callSeq: AtomicInteger = AtomicInteger(0)
    private var permissionRequestInFlight: Boolean = false

    fun register(engine: FlutterEngine) {
        MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL)
            .setMethodCallHandler(this)
    }

    override fun onMethodCall(
        call: io.flutter.plugin.common.MethodCall,
        result: MethodChannel.Result,
    ) {
        if (call.method == "ensurePermission") {
            result.success(mapOf("granted" to hasPermission()))
            return
        }
        // Calendar listing & todo-calendar selection never need the
        // runtime permission — they only need the system to know
        // about the user's calendars. Pass them through directly.
        if (call.method == "listCalendars" ||
            call.method == "setTodoCalendar" ||
            call.method == "getTodoCalendar"
        ) {
            try {
                handle(call.method, call.arguments as? Map<String, Any?>, result)
            } catch (e: Exception) {
                result.error("BRIDGE_ERROR", e.message, null)
            }
            return
        }
        if (!hasPermission()) {
            if (currentStatus() == "permanently_denied") {
                result.error(
                    "PERMANENTLY_DENIED",
                    "Calendar permission permanently denied; open system settings to enable it",
                    null,
                )
                return
            }
            val callId = nextCallId()
            pendingResults[callId] = Pending(
                result = result,
                method = call.method,
                args = call.arguments as? Map<String, Any?>,
                delivered = AtomicBoolean(false),
            )
            requestPermissionIfNeeded()
            return
        }
        try {
            handle(call.method, call.arguments as? Map<String, Any?>, result)
        } catch (e: SecurityException) {
            result.error("PERMISSION_DENIED", e.message, null)
        } catch (e: Exception) {
            result.error("BRIDGE_ERROR", e.message, null)
        }
    }

    private fun handle(
        method: String,
        args: Map<String, Any?>?,
        result: MethodChannel.Result,
    ) {
        when (method) {
            "listCalendars" -> {
                result.success(listWritableCalendars())
            }
            "setTodoCalendar" -> {
                val id = (args?.get("id") as? String)?.toLongOrNull()
                    ?: throw IllegalArgumentException("id (long) required")
                prefs.edit().putLong(PREF_TODO_CALENDAR, id).apply()
                result.success(mapOf("ok" to true))
            }
            "getTodoCalendar" -> {
                val id = prefs.getLong(PREF_TODO_CALENDAR, -1L)
                result.success(if (id < 0) null else id.toString())
            }
            "listReminders" -> {
                val includeCompleted = (args?.get("includeCompleted") as? Boolean) ?: false
                val max = (args?.get("max") as? Number)?.toInt() ?: 50
                val list = listReminders(includeCompleted, max)
                result.success(list)
            }
            "createReminder" -> {
                val calId = todoCalendarIdOrThrow()
                val r = createReminder(calId, args)
                result.success(r)
            }
            "completeReminder" -> {
                val id = (args?.get("id") as? String)?.toLongOrNull()
                    ?: throw IllegalArgumentException("id (long) required")
                val r = completeReminder(id)
                if (r == null) result.success(null) else result.success(r)
            }
            "updateReminder" -> {
                val id = (args?.get("id") as? String)?.toLongOrNull()
                    ?: throw IllegalArgumentException("id (long) required")
                val r = updateReminder(id, args!!)
                if (r == null) result.success(null) else result.success(r)
            }
            "deleteReminder" -> {
                val id = (args?.get("id") as? String)?.toLongOrNull()
                    ?: throw IllegalArgumentException("id (long) required")
                val ok = deleteReminder(id)
                result.success(mapOf("ok" to ok))
            }
            else -> result.notImplemented()
        }
    }

    // -------- Permission --------

    private fun hasPermission(): Boolean {
        val read = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.READ_CALENDAR,
        ) == PackageManager.PERMISSION_GRANTED
        if (!read) return false
        return ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.WRITE_CALENDAR,
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun currentStatus(): String {
        if (hasPermission()) return "granted"
        val rationaleOff = !ActivityCompat.shouldShowRequestPermissionRationale(
            activity,
            Manifest.permission.READ_CALENDAR,
        )
        return if (rationaleOff && hasEverRequested()) "permanently_denied" else "denied"
    }

    private fun hasEverRequested(): Boolean =
        prefs.getBoolean("reminders_permission_requested", false)

    private fun markPermissionRequested() {
        prefs.edit().putBoolean("reminders_permission_requested", true).apply()
    }

    private fun requestPermissionIfNeeded() {
        if (permissionRequestInFlight) return
        permissionRequestInFlight = true
        markPermissionRequested()
        ActivityCompat.requestPermissions(
            activity,
            arrayOf(
                Manifest.permission.READ_CALENDAR,
                Manifest.permission.WRITE_CALENDAR,
            ),
            PERMISSION_REQUEST_CODE,
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray,
    ): Boolean {
        if (requestCode != PERMISSION_REQUEST_CODE) return false
        permissionRequestInFlight = false
        val snapshot = pendingResults.toMap()
        pendingResults.clear()

        if (!hasPermission()) {
            val code = if (currentStatus() == "permanently_denied") {
                "PERMANENTLY_DENIED"
            } else {
                "PERMISSION_DENIED"
            }
            val msg = "Calendar permission was denied; please grant it in system settings."
            for ((_, p) in snapshot) {
                if (p.delivered.compareAndSet(false, true)) {
                    p.result.error(code, msg, null)
                }
            }
            return true
        }
        for ((_, p) in snapshot) {
            if (!p.delivered.compareAndSet(false, true)) continue
            try {
                handle(p.method, p.args, p.result)
            } catch (e: Exception) {
                p.result.error("BRIDGE_ERROR", e.message, null)
            }
        }
        return true
    }

    private fun nextCallId(): String = "rem-${callSeq.incrementAndGet()}"

    // -------- Calendar selection --------

    private fun listWritableCalendars(): List<Map<String, Any?>> {
        val projection = arrayOf(
            CalendarContract.Calendars._ID,
            CalendarContract.Calendars.CALENDAR_DISPLAY_NAME,
            CalendarContract.Calendars.ACCOUNT_NAME,
            CalendarContract.Calendars.CALENDAR_ACCESS_LEVEL,
        )
        val out = mutableListOf<Map<String, Any?>>()
        context.contentResolver.query(
            CalendarContract.Calendars.CONTENT_URI,
            projection,
            "${CalendarContract.Calendars.CALENDAR_ACCESS_LEVEL} >= ${CalendarContract.Calendars.CAL_ACCESS_CONTRIBUTOR}",
            null,
            null,
        )?.use { c ->
            while (c.moveToNext()) {
                out.add(
                    mapOf(
                        "id" to c.getLong(
                            c.getColumnIndexOrThrow(CalendarContract.Calendars._ID)
                        ).toString(),
                        "displayName" to c.getString(
                            c.getColumnIndexOrThrow(
                                CalendarContract.Calendars.CALENDAR_DISPLAY_NAME
                            )
                        ),
                        "accountName" to c.getString(
                            c.getColumnIndexOrThrow(CalendarContract.Calendars.ACCOUNT_NAME)
                        ),
                    )
                )
            }
        }
        return out
    }

    private fun todoCalendarIdOrThrow(): Long {
        val id = prefs.getLong(PREF_TODO_CALENDAR, -1L)
        if (id < 0) {
            throw NoTodoCalendarException()
        }
        return id
    }

    class NoTodoCalendarException :
        Exception("Please pick a 'todo' calendar in Agent Buddy settings before using reminders")

    // -------- Reminder CRUD (backed by all-day events) --------

    private fun listReminders(includeCompleted: Boolean, max: Int): List<Map<String, Any?>> {
        val calId = prefs.getLong(PREF_TODO_CALENDAR, -1L)
        if (calId < 0) throw NoTodoCalendarException()

        val projection = arrayOf(
            CalendarContract.Events._ID,
            CalendarContract.Events.TITLE,
            CalendarContract.Events.DESCRIPTION,
            CalendarContract.Events.DTSTART,
            CalendarContract.Events.ALL_DAY,
            CalendarContract.Events.CALENDAR_DISPLAY_NAME,
        )
        // "Completed" = due-date strictly in the past. List only
        // events we own (description starts with the marker).
        val nowMs = System.currentTimeMillis()
        val (selection, args) = if (includeCompleted) {
            "${CalendarContract.Events.CALENDAR_ID} = ? " +
                "AND ${CalendarContract.Events.ALL_DAY} = 1 " +
                "AND ${CalendarContract.Events.DESCRIPTION} LIKE ?" to
                arrayOf(calId.toString(), "$MARKER%")
        } else {
            "${CalendarContract.Events.CALENDAR_ID} = ? " +
                "AND ${CalendarContract.Events.ALL_DAY} = 1 " +
                "AND ${CalendarContract.Events.DTSTART} >= ? " +
                "AND ${CalendarContract.Events.DESCRIPTION} LIKE ?" to
                arrayOf(calId.toString(), nowMs.toString(), "$MARKER%")
        }
        val out = mutableListOf<Map<String, Any?>>()
        context.contentResolver.query(
            CalendarContract.Events.CONTENT_URI,
            projection,
            selection,
            args,
            "${CalendarContract.Events.DTSTART} ASC LIMIT $max",
        )?.use { c ->
            val idIdx = c.getColumnIndexOrThrow(CalendarContract.Events._ID)
            val titleIdx = c.getColumnIndexOrThrow(CalendarContract.Events.TITLE)
            val descIdx = c.getColumnIndexOrThrow(CalendarContract.Events.DESCRIPTION)
            val startIdx = c.getColumnIndexOrThrow(CalendarContract.Events.DTSTART)
            val calNameIdx = c.getColumnIndexOrThrow(
                CalendarContract.Events.CALENDAR_DISPLAY_NAME
            )
            while (c.moveToNext()) {
                val id = c.getLong(idIdx)
                val dueMs = c.getLong(startIdx)
                val isCompleted = dueMs < nowMs
                out.add(
                    mapOf<String, Any?>(
                        "id" to id.toString(),
                        "title" to (c.getString(titleIdx) ?: ""),
                        "notes" to stripMarker(c.getString(descIdx)),
                        "dueMs" to dueMs,
                        "completed" to isCompleted,
                        "completedAtMs" to if (isCompleted) dueMs else null,
                        "calendarName" to c.getString(calNameIdx),
                    )
                )
            }
        }
        return out
    }

    private fun createReminder(
        calId: Long,
        args: Map<String, Any?>?,
    ): Map<String, Any?> {
        val title = args?.get("title") as? String
            ?: throw IllegalArgumentException("title required")
        val notes = args?.get("notes") as? String
        val dueMs = (args?.get("dueMs") as? Number)?.toLong()
            ?: startOfDayUtcMs(System.currentTimeMillis() + 24L * 60 * 60 * 1000)
        val cv = ContentValues().apply {
            put(CalendarContract.Events.CALENDAR_ID, calId)
            put(CalendarContract.Events.TITLE, title)
            // Prepend the marker so the row is identifiable as an
            // Agent Buddy reminder. We can't rely on the
            // `categories` column — some OEM providers (Huawei /
            // Xiaomi local calendars) reject it with
            // `IllegalArgumentException: column 'categories' is
            // invalid`.
            put(CalendarContract.Events.DESCRIPTION, prependMarker(notes))
            put(CalendarContract.Events.DTSTART, dueMs)
            put(CalendarContract.Events.DTEND, dueMs)
            put(CalendarContract.Events.ALL_DAY, 1)
            put(CalendarContract.Events.EVENT_TIMEZONE, "UTC")
        }
        val uri = context.contentResolver.insert(CalendarContract.Events.CONTENT_URI, cv)
            ?: throw IllegalStateException("failed to insert reminder")
        val id = ContentUris.parseId(uri)
        val now = System.currentTimeMillis()
        val isCompleted = dueMs < now
        return mapOf<String, Any?>(
            "id" to id.toString(),
            "title" to title,
            "notes" to notes,
            "dueMs" to dueMs,
            "completed" to isCompleted,
            "completedAtMs" to if (isCompleted) dueMs else null,
        )
    }

    private fun completeReminder(id: Long): Map<String, Any?>? {
        // "Complete" = move due date to the past so it stops showing
        // up in the default (non-completed) list.
        val past = startOfDayUtcMs(System.currentTimeMillis() - 24L * 60 * 60 * 1000)
        val cv = ContentValues().apply {
            put(CalendarContract.Events.DTSTART, past)
            put(CalendarContract.Events.DTEND, past)
        }
        val uri = ContentUris.withAppendedId(CalendarContract.Events.CONTENT_URI, id)
        val rows = context.contentResolver.update(uri, cv, null, null)
        if (rows == 0) return null
        val existing = readReminder(id) ?: return null
        return existing.toMutableMap().apply {
            put("dueMs", past)
            put("completed", true)
            put("completedAtMs", past)
        }
    }

    private fun updateReminder(id: Long, args: Map<String, Any?>): Map<String, Any?>? {
        val cv = ContentValues().apply {
            (args["title"] as? String)?.let { put(CalendarContract.Events.TITLE, it) }
            (args["notes"] as? String)?.let {
                put(CalendarContract.Events.DESCRIPTION, prependMarker(it))
            }
            (args["dueMs"] as? Number)?.toLong()?.let {
                put(CalendarContract.Events.DTSTART, it)
                put(CalendarContract.Events.DTEND, it)
            }
        }
        if (cv.size() == 0) return readReminder(id)
        val uri = ContentUris.withAppendedId(CalendarContract.Events.CONTENT_URI, id)
        val rows = context.contentResolver.update(uri, cv, null, null)
        if (rows == 0) return null
        return readReminder(id)
    }

    private fun deleteReminder(id: Long): Boolean {
        val uri = ContentUris.withAppendedId(CalendarContract.Events.CONTENT_URI, id)
        return try {
            context.contentResolver.delete(uri, null, null) > 0
        } catch (e: Exception) {
            false
        }
    }

    private fun readReminder(id: Long): Map<String, Any?>? {
        val uri = ContentUris.withAppendedId(CalendarContract.Events.CONTENT_URI, id)
        context.contentResolver.query(uri, null, null, null, null)?.use { c ->
            if (c.moveToFirst()) {
                val title = c.getString(
                    c.getColumnIndexOrThrow(CalendarContract.Events.TITLE)
                ) ?: ""
                val desc = c.getString(
                    c.getColumnIndexOrThrow(CalendarContract.Events.DESCRIPTION)
                )
                val dueMs = c.getLong(
                    c.getColumnIndexOrThrow(CalendarContract.Events.DTSTART)
                )
                val now = System.currentTimeMillis()
                val isCompleted = dueMs < now
                return mapOf<String, Any?>(
                    "id" to id.toString(),
                    "title" to title,
                    "notes" to stripMarker(desc),
                    "dueMs" to dueMs,
                    "completed" to isCompleted,
                    "completedAtMs" to if (isCompleted) dueMs else null,
                )
            }
        }
        return null
    }

    // -------- Marker helpers --------

    private fun prependMarker(notes: String?): String =
        if (notes.isNullOrEmpty()) MARKER else "$MARKER$notes"

    private fun stripMarker(desc: String?): String? {
        if (desc.isNullOrEmpty()) return desc
        return if (desc.startsWith(MARKER)) desc.substring(MARKER.length) else desc
    }

    private fun startOfDayUtcMs(timestampMs: Long): Long {
        val cal = java.util.Calendar.getInstance(java.util.TimeZone.getTimeZone("UTC"))
        cal.timeInMillis = timestampMs
        cal.set(java.util.Calendar.HOUR_OF_DAY, 0)
        cal.set(java.util.Calendar.MINUTE, 0)
        cal.set(java.util.Calendar.SECOND, 0)
        cal.set(java.util.Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }
}
